# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
path = os.getcwd()
print(path)

C:\Users\j9903\Desktop\SpringBoard\Assignments\Capstone Project 1


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df0 = pd.read_csv(path + '\\data\\step3_output.csv')

In [4]:
df0.columns

Index(['Unnamed: 0', 'Name', 'state', 'summit_elev', 'vertical_drop', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters'],
      dtype='object')

In [5]:
df0 = df0.drop('Unnamed: 0', axis=1)

In [6]:
df0.describe()

summit_elev  vertical_drop  trams  fastEight  fastSixes   fastQuads  \
count    151.000000     151.000000  151.0      151.0      151.0  151.000000   
mean    3695.377483     849.814570    0.0        0.0        0.0    0.225166   
std     3246.874203     572.536792    0.0        0.0        0.0    0.567711   
min      315.000000     175.000000    0.0        0.0        0.0    0.000000   
25%     1260.000000     400.000000    0.0        0.0        0.0    0.000000   
50%     2000.000000     700.000000    0.0        0.0        0.0    0.000000   
75%     6144.500000    1150.000000    0.0        0.0        0.0    0.000000   
max    12075.000000    2590.000000    0.0        0.0        0.0    2.000000   

             quad      triple      double     surface  ...  SkiableTerrain_ac  \
count  151.000000  151.000000  151.000000  151.000000  ...         151.000000   
mean     0.523179    1.132450    1.675497    2.165563  ...         251.368228   
std      0.710253    1.192619    1.393077    1.387710  ...         323.518663   
min      0.000000    0.000000    0.000000    0.000000  ...           8.000000   
25%      0.000000    0.000000    1.000000    1.000000  ...          60.000000   
50%      0.000000    1.000000    1.000000    2.000000  ...         125.000000   
75%      1.000000    2.000000    2.000000    3.000000  ...         272.500000   
max      2.000000    5.000000    6.000000    6.000000  ...        1704.000000   

       Snow Making_ac  daysOpenLastYear   yearsOpen  averageSnowfall  \
count      151.000000        151.000000  151.000000       151.000000   
mean       103.943475         92.529801   58.172185       145.964456   
std         68.983952         39.014324   14.861477       108.414187   
min          2.000000          0.000000   22.000000        18.000000   
25%         45.000000         75.000000   50.000000        53.000000   
50%         86.000000        100.000000   57.000000       110.000000   
75%        174.873239        118.000000   67.000000       210.000000   
max        300.000000        175.000000   95.000000       500.000000   

       AdultWeekday  AdultWeekend  projectedDaysOpen  NightSkiing_ac  \
count    151.000000    151.000000         151.000000      151.000000   
mean      50.264426     57.138247         108.245033       41.496689   
std       15.596557     15.262982          26.842495       45.079763   
min       20.000000     20.000000          30.000000        0.000000   
25%       40.000000     46.000000          93.000000        0.000000   
50%       49.000000     57.000000         107.000000       35.000000   
75%       57.916957     65.000000         124.000000       69.500000   
max       89.000000     93.000000         200.000000      180.000000   

         clusters  
count  151.000000  
mean     1.377483  
std      0.780954  
min      0.000000  
25%      1.000000  
50%      2.000000  
75%      2.000000  
max      2.000000  

[8 rows x 24 columns]

In [7]:
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151 entries, 0 to 150
Data columns (total 26 columns):
Name                 151 non-null object
state                151 non-null object
summit_elev          151 non-null int64
vertical_drop        151 non-null int64
trams                151 non-null int64
fastEight            151 non-null float64
fastSixes            151 non-null int64
fastQuads            151 non-null int64
quad                 151 non-null int64
triple               151 non-null int64
double               151 non-null int64
surface              151 non-null int64
total_chairs         151 non-null int64
Runs                 151 non-null float64
TerrainParks         151 non-null float64
LongestRun_mi        151 non-null float64
SkiableTerrain_ac    151 non-null float64
Snow Making_ac       151 non-null float64
daysOpenLastYear     151 non-null float64
yearsOpen            151 non-null float64
averageSnowfall      151 non-null float64
AdultWeekday         151 non-null f

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [8]:
df = pd.concat([df0.drop('state', axis=1), pd.get_dummies(df0["state"])], axis=1)

In [9]:
df.head(10)

Name  summit_elev  vertical_drop  trams  \
0            Eaglecrest Ski Area         2600           1540      0   
1               Hilltop Ski Area         2090            294      0   
2            Sunrise Park Resort        11100           1800      0   
3  Yosemite Ski & Snowboard Area         7800            600      0   
4               Donner Ski Ranch         8012            750      0   
5                  June Mountain        10090           2590      0   
6            Mt. Shasta Ski Park         6890           1435      0   
7                  Mountain High         8200           1600      0   
8                      Mt. Baldy         8600           2100      0   
9                 Ski China Peak         8709           1679      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       0       4  ...             0   
1        0.0          0          0     0       1       0  ...             0   
2        0.0          0          1     2       3       1  ...             0   
3        0.0          0          0     0       1       3  ...             0   
4        0.0          0          0     0       1       5  ...             0   
5        0.0          0          2     0       0       4  ...             0   
6        0.0          0          0     0       3       0  ...             0   
7        0.0          0          2     2       2       5  ...             0   
8        0.0          0          0     0       0       4  ...             0   
9        0.0          0          0     1       4       2  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   
5             0          0     0        0         0           0   
6             0          0     0        0         0           0   
7             0          0     0        0         0           0   
8             0          0     0        0         0           0   
9             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  
5              0          0        0  
6              0          0        0  
7              0          0        0  
8              0          0        0  
9              0          0        0  

[10 rows x 59 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [10]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled = scaler.transform(X) 

In [11]:
X.shape

(151, 57)

In [12]:
X_scaled.shape

(151, 57)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [13]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [14]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [15]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [16]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
explained_variance_score(y_test,y_pred)

-1.1944895906874592e+22

In [17]:
mean_absolute_error(y_test,y_pred)

271703868443.54605

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [18]:
lm.intercept_

68375808008.70406

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [19]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Coefficient
trams              6.359937e+13
fastSixes          4.085425e+13
total_chairs       1.527862e+13
fastEight          1.402470e+13
double             8.658934e+12
surface            8.625578e+12
triple             7.412951e+12
quad               4.414713e+12
fastQuads          3.528716e+12
New York           2.994770e+12
Michigan           2.792783e+12
Pennsylvania       2.683488e+12
Wisconsin          2.567727e+12
New Hampshire      2.567727e+12
California         2.444582e+12
Minnesota          2.312875e+12
New Mexico         2.016967e+12
Vermont            2.016967e+12
Colorado           1.847569e+12
Connecticut        1.847569e+12
Maine              1.847569e+12
Idaho              1.658166e+12
Washington         1.658166e+12
North Carolina     1.658166e+12
Wyoming            1.440890e+12
Oregon             1.440890e+12
Utah               1.440890e+12
Ohio               1.440890e+12
Illinois           1.440890e+12
West Virginia      1.440890e+12
Massachusetts      1.440890e+12
Montana            1.180449e+12
Nevada             1.180449e+12
Indiana            1.180449e+12
Alaska             1.180449e+12
Virginia           1.180449e+12
South Dakota       8.375001e+11
Arizona            8.375001e+11
Missouri           8.375001e+11
Maryland           8.375001e+11
Rhode Island       8.375001e+11
Tennessee          8.375001e+11
Iowa               7.088456e+07
AdultWeekday       9.954590e+00
summit_elev        3.107335e+00
clusters           1.682861e+00
yearsOpen          1.480469e+00
Runs               1.022705e+00
averageSnowfall    9.667969e-01
projectedDaysOpen  9.023438e-01
NightSkiing_ac     7.441406e-01
LongestRun_mi      6.328125e-01
SkiableTerrain_ac  3.740234e-01
daysOpenLastYear   2.602539e-01
Snow Making_ac     2.441406e-01
TerrainParks       1.826172e-01
vertical_drop      4.248941e-03

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [20]:
df_m2 = df0.copy()

In [21]:
X_m2 = df_m2.drop(['Name','state','AdultWeekend'], axis=1)
y_m2 = df_m2.AdultWeekend
scaler_m2 = preprocessing.StandardScaler().fit(X_m2)
X_scaled_m2 =scaler_m2.transform(X_m2)
y_m2 = y_m2.ravel()
X_train_m2, X_test_m2, y_train_m2, y_test_m2 = train_test_split(X_scaled_m2, y_m2, test_size=0.25, random_state=1)

In [22]:
lm_m2 = linear_model.LinearRegression()
model_m2 = lm_m2.fit(X_train_m2,y_train_m2)
y_pred_m2 = model_m2.predict(X_test_m2)

In [23]:
print("explained_variance_score = " + str(explained_variance_score(y_test_m2,y_pred_m2)))
print("mean_absolute_error = " + str(mean_absolute_error(y_test_m2,y_pred_m2)))
print("Intercept of the model = " + str(lm_m2.intercept_))

explained_variance_score = 0.712592093141009
mean_absolute_error = 6.254795421460555
Intercept of the model = 57.430797573237186


In [26]:
pd.DataFrame(abs(lm_m2.coef_), X_m2.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Coefficient
AdultWeekday       1.116559e+01
summit_elev        3.265363e+00
clusters           2.989427e+00
averageSnowfall    1.594709e+00
Runs               1.330914e+00
triple             1.073047e+00
quad               1.063912e+00
daysOpenLastYear   1.063812e+00
LongestRun_mi      7.814994e-01
TerrainParks       7.497670e-01
vertical_drop      7.325549e-01
yearsOpen          7.051949e-01
surface            6.972185e-01
total_chairs       6.024067e-01
projectedDaysOpen  5.079834e-01
double             4.421399e-01
NightSkiing_ac     4.150987e-01
fastQuads          3.576134e-01
Snow Making_ac     1.331402e-01
SkiableTerrain_ac  1.020249e-01
fastEight          1.776357e-15
trams              8.881784e-16
fastSixes          0.000000e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [27]:
df_m3 = df0.copy()

In [29]:
X_m3 = df_m3.drop(['Name','state','AdultWeekend','summit_elev'], axis=1)
y_m3 = df_m3.AdultWeekend
scaler_m3 = preprocessing.StandardScaler().fit(X_m3)
X_scaled_m3 =scaler_m3.transform(X_m3)
y_m3 = y_m3.ravel()
X_train_m3, X_test_m3, y_train_m3, y_test_m3 = train_test_split(X_scaled_m3, y_m3, test_size=0.25, random_state=1)

In [30]:
lm_m3 = linear_model.LinearRegression()
model_m3 = lm_m3.fit(X_train_m3,y_train_m3)
y_pred_m3 = model_m3.predict(X_test_m3)

In [31]:
print("explained_variance_score = " + str(explained_variance_score(y_test_m3,y_pred_m3)))
print("mean_absolute_error = " + str(mean_absolute_error(y_test_m3,y_pred_m3)))
print("Intercept of the model = " + str(lm_m3.intercept_))

explained_variance_score = 0.6971188758658406
mean_absolute_error = 6.350158930805728
Intercept of the model = 57.29229169337912


In [32]:
pd.DataFrame(abs(lm_m3.coef_), X_m3.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Coefficient
AdultWeekday       1.067471e+01
clusters           2.574753e+00
averageSnowfall    2.512312e+00
Runs               1.331457e+00
triple             1.167222e+00
yearsOpen          8.987897e-01
projectedDaysOpen  8.883705e-01
quad               8.535625e-01
total_chairs       8.510341e-01
double             8.420608e-01
NightSkiing_ac     8.052314e-01
daysOpenLastYear   6.416391e-01
surface            5.934652e-01
TerrainParks       5.627604e-01
LongestRun_mi      5.379993e-01
fastQuads          4.507402e-01
SkiableTerrain_ac  2.976596e-01
vertical_drop      1.432614e-01
Snow Making_ac     2.641446e-02
trams              8.881784e-16
fastEight          8.881784e-16
fastSixes          4.440892e-16

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.1944895906874592e+22 | 271703868443.54605 |'base_elev'|
| Model 2. | 0.712592093141009| 6.254795421460555 |'state','base_elev'|
| Model 3. | 0.6971188758658406 | 6.350158930805728 |'state','summit_elev','base_elev'|

Model Selection: Model 2

The success criteria of the model performance is 
1. MAE > 40
2. Lower EV (for EV, the best possible score is 1.0 and lower values are worse)
Therefore model 1 is not selected due to its MAE value. Model 3 is not selected because it EV amd MSE is worse than model 2.